File: lib/ingest/legacy_xref.ex
# lib/ingest/legacy_xref.ex — the durable legacy -> new cross-reference (gr-0c2).
#
# Stage 4b of the legacy-medipim ingest, a fold over the re-derivation output (gr-chq:
# `Rederivation.run/2` / `from_claims/2` -> `%{log, ledger, clusters, shared}`). Per the design
# (docs/plans/2026-06-05-legacy-history-ingest-design.md, "LegacyXref — the durable legacy -> new
# map"): the legacy `entity` was NEVER a clustering input — it rode along only as the :grouping
# claims ClaimMapping synthesized (one per listing-code, `%{code, product: legacy_entity}`). Here
# we read those grouping claims back out and join them against the re-derived surrogate keys to
# answer the migration question: where did every existing legacy `entity` id land, and did it stay
# put (stable), get absorbed with others (merged), or fragment across keys (split)?
#
# OUTPUT — two folds plus a resolver:
#
#   key_to_legacy:  SK   -> sorted-unique [legacy_entity, ...]   (provenance of each surrogate key)
#   legacy_to_key:  ent  -> %{primary: SK, all: [SK, ...], relation: :stable | :split |
#                            {:merged, [other_ent, ...]}}
#   resolve_legacy(xref, ent) -> {:ok, primary, status} | {:error, :unknown_legacy}
#
# `relation` IS the migration diff classification (confirm / merge / split / collision) the design
# promises "for free" — but the diff VIEW / rendering is gr-swc, and the golden-record projection
# (Catalog/Api output) is gr-8r6. This module produces ONLY the maps + resolver; it does not render
# a diff and does not project records.
#
# SPLIT PRIMARY — the engine's split "keep" heuristic, re-implemented here to the design's spec
# (CNK ▸ GTIN spine, then most listings, then lowest key). NOTE: the engine's IdentityLedger has a
# private `has_spine?/1` that only checks `:gtin`; the design wants CNK to outrank GTIN, so we
# compute spine rank ourselves rather than borrow it. "Most listings" = the count of DISTINCT
# grouping-claim sources contributing codes to that key — i.e. how many real org listings the
# surrogate key is backed by (a listing = a (legacy_entity, source) pair, and ClaimMapping stamps
# each grouping claim with its listing `source`).

defmodule LegacyXref do
  @doc """
  Build the cross-reference from a re-derivation result `%{log: log, ledger: ledger}` (the map
  returned by `Rederivation.run/2` / `Rederivation.from_claims/2`; extra keys like `clusters` /
  `shared` are ignored).

  Returns `%{key_to_legacy: %{SK => [legacy_entity]}, legacy_to_key: %{legacy_entity => %{primary,
  all, relation}}}`. See the moduledoc for the shapes and the `relation` taxonomy.
  """
  def build(%{log: log, ledger: ledger}) do
    groupings = groupings(log)

    # legacy_entity -> MapSet of its canonical {scheme, code} (drawn from its grouping claims). Used
    # by the over-merge guard to find the bridge — the codes shared across the merged entities.
    entity_codes =
      Enum.reduce(groupings, %{}, fn g, acc ->
        Map.update(acc, g.data.product, MapSet.new([g.data.code]), &MapSet.put(&1, g.data.code))
      end)

    # SK -> %{codes: MapSet, sources: MapSet} drawn from the grouping claims landing on its members.
    per_key =
      for {key, member_codes} <- ledger.members, into: %{} do
        contributing =
          Enum.filter(groupings, fn g -> MapSet.member?(member_codes, g.data.code) end)

        {key,
         %{
           entities: contributing |> Enum.map(& &1.data.product) |> Enum.sort() |> Enum.uniq(),
           sources: contributing |> Enum.map(& &1.source) |> MapSet.new(),
           codes: member_codes
         }}
      end

    key_to_legacy = Map.new(per_key, fn {key, info} -> {key, info.entities} end)
    legacy_to_key = invert(key_to_legacy, per_key, entity_codes)

    %{key_to_legacy: key_to_legacy, legacy_to_key: legacy_to_key}
  end

  @doc """
  Convenience: run `Rederivation.run(envelopes, at)` internally, then `build/1` its result. For
  callers that only want the xref and have raw `%HistoryEnvelope{}`s.
  """
  def from_envelopes(envelopes, at) when is_list(envelopes) do
    envelopes |> Rederivation.run(at) |> build()
  end

  @doc """
  Resolve a legacy `entity` id to its surrogate key, disclosing alternates:

    * stable  -> `{:ok, SK, :stable}`
    * split   -> `{:ok, primary_SK, {:split, [all_SKs...]}}`  (answer with the primary, list all)
    * merged  -> `{:ok, SK, {:merged, [other_legacy_ids...]}}` (the co-tenants on that key)
    * merged (suspect) -> `{:ok, SK, {:merged, [other_legacy_ids...], :suspect}}` (gr-ose: the
                  co-tenants were bridged SOLELY by a reusable barcode/GS1 code — an over-merge the
                  migration diff must surface as needs-review; the 3-tuple passes through unchanged)

  Unknown id -> `{:error, :unknown_legacy}`. Mirrors how a stale CNK redirects today: callers get a
  single authoritative key plus enough context to see what else it now relates to.
  """
  def resolve_legacy(%{legacy_to_key: legacy_to_key}, legacy_id) do
    case Map.fetch(legacy_to_key, legacy_id) do
      {:ok, %{primary: primary, all: all, relation: :split}} -> {:ok, primary, {:split, all}}
      {:ok, %{primary: primary, relation: relation}} -> {:ok, primary, relation}
      :error -> {:error, :unknown_legacy}
    end
  end

  # ── internals ───────────────────────────────────────────────────────────────

  defp groupings(log) do
    Enum.filter(log, &match?(%Events.ClaimAsserted{kind: :grouping}, &1))
  end

  # legacy_entity -> %{primary, all, relation}. `all` = every SK the entity's grouping claims point
  # at (sorted). `relation` is decided from that placement plus each key's co-tenancy.
  defp invert(key_to_legacy, per_key, entity_codes) do
    entity_to_keys =
      Enum.reduce(key_to_legacy, %{}, fn {key, entities}, acc ->
        Enum.reduce(entities, acc, fn e, acc -> Map.update(acc, e, [key], &[key | &1]) end)
      end)

    Map.new(entity_to_keys, fn {entity, keys} ->
      all = keys |> Enum.uniq() |> Enum.sort_by(&key_num/1)

      {entity,
       %{
         primary: primary(all, entity, key_to_legacy, per_key),
         all: all,
         relation: relation(all, entity, key_to_legacy, entity_codes)
       }}
    end)
  end

  # PRECEDENCE: split outranks merged. A legacy entity spread over >1 key has FRAGMENTED — that is
  # the dominant fact a migrator must act on, regardless of whether any one of those keys also holds
  # OTHER legacy entities. (A degenerate case — split AND a co-tenant on one of the keys — is
  # therefore reported as :split; the co-tenancy still surfaces via key_to_legacy for that key.)
  defp relation([_single = key], entity, key_to_legacy, entity_codes) do
    case Enum.sort(Enum.reject(key_to_legacy[key], &(&1 == entity))) do
      [] -> :stable
      others -> merged(entity, others, entity_codes)
    end
  end

  defp relation(_many, _entity, _key_to_legacy, _entity_codes), do: :split

  # OVER-MERGE GUARD (gr-ose): the merge IS applied (codes won, one SK) — we only TAG it. The bridge
  # is the set of codes shared between `entity` and ANY co-tenant on this key. If any shared bridge
  # code is national-grade, the merge is trusted: `{:merged, others}`. If the entities are bridged
  # SOLELY by barcode/GS1 codes, tag it suspect: `{:merged, others, :suspect}`. A national share
  # always wins (it is the re-derivation working as intended), so a national bridge alone keeps the
  # plain 2-tuple even if a barcode is also shared.
  defp merged(entity, others, entity_codes) do
    mine = Map.get(entity_codes, entity, MapSet.new())

    bridge =
      Enum.reduce(others, MapSet.new(), fn other, acc ->
        shared = MapSet.intersection(mine, Map.get(entity_codes, other, MapSet.new()))
        MapSet.union(acc, shared)
      end)

    if Enum.any?(bridge, fn {scheme, _code} -> CodeRegistry.national_grade?(scheme) end) do
      {:merged, others}
    else
      {:merged, others, :suspect}
    end
  end

  # Non-split: the sole key. Split: the engine "keep" heuristic per the design — spine (CNK ▸ GTIN),
  # then most listings (distinct grouping-claim sources backing the key), then lowest key id.
  defp primary([single], _entity, _key_to_legacy, _per_key), do: single

  defp primary(all, _entity, _key_to_legacy, per_key) do
    Enum.max_by(all, fn key ->
      info = per_key[key]
      {spine_rank(info.codes), MapSet.size(info.sources), -key_num(key)}
    end)
  end

  # CNK is a stronger identity spine than GTIN (the design's CNK ▸ GTIN ordering). 2 = has a CNK,
  # 1 = has a GTIN but no CNK, 0 = neither.
  defp spine_rank(codes) do
    cond do
      Enum.any?(codes, &match?({:cnk, _}, &1)) -> 2
      Enum.any?(codes, &match?({:gtin, _}, &1)) -> 1
      true -> 0
    end
  end

  defp key_num("SK_" <> n), do: String.to_integer(n)
end

File: php/src/LegacyXref.php
<?php

declare(strict_types=1);

namespace Ingot;

/**
 * The durable legacy → new cross-reference — ported from `LegacyXref` (lib/ingest/legacy_xref.ex).
 *
 * Reads the grouping claims back out and joins them against the re-derived surrogate keys to answer:
 * where did each legacy entity land, and did it stay put (stable), get absorbed (merged), or
 * fragment (split)? The over-merge guard tags a barcode-only bridge as :suspect.
 */
final class LegacyXref
{
    /**
     * @param array{log: list<array<string,mixed>>, ledger: LedgerState} $rederivation
     * @return array{key_to_legacy: array<string, list<mixed>>, legacy_to_key: array<string, array<string,mixed>>}
     */
    public static function build(array $rederivation): array
    {
        $log = $rederivation['log'];
        $ledger = $rederivation['ledger'];
        $groupings = self::groupings($log);

        // entity => code-set of its grouping codes.
        $entityCodes = [];
        foreach ($groupings as $g) {
            $e = self::entityKey($g['data']['product']);
            if (!isset($entityCodes[$e])) {
                $entityCodes[$e] = ['value' => $g['data']['product'], 'codes' => []];
            }
            $entityCodes[$e]['codes'][Codes::key($g['data']['code'])] = $g['data']['code'];
        }

        // SK => {entities, sources, codes} — product lane only.
        $perKey = [];
        foreach (Lanes::partitionMembers($ledger->members)['product'] as $key => $memberCodes) {
            $contributing = [];
            foreach ($groupings as $g) {
                if (Sets::member($memberCodes, $g['data']['code'])) {
                    $contributing[] = $g;
                }
            }
            $entities = [];
            $entSeen = [];
            $sources = [];
            foreach ($contributing as $g) {
                $ek = self::entityKey($g['data']['product']);
                if (!isset($entSeen[$ek])) {
                    $entSeen[$ek] = true;
                    $entities[] = $g['data']['product'];
                }
                $sources[$g['source']] = true;
            }
            usort($entities, self::compareEntities(...));
            $perKey[$key] = ['entities' => $entities, 'sources' => $sources, 'codes' => $memberCodes];
        }

        $keyToLegacy = [];
        foreach ($perKey as $key => $info) {
            $keyToLegacy[$key] = $info['entities'];
        }

        $legacyToKey = self::invert($keyToLegacy, $perKey, $entityCodes);

        return ['key_to_legacy' => $keyToLegacy, 'legacy_to_key' => $legacyToKey];
    }

    /**
     * @param list<array<string,mixed>> $envelopes
     * @return array{key_to_legacy: array<string, list<mixed>>, legacy_to_key: array<string, array<string,mixed>>}
     */
    public static function fromEnvelopes(array $envelopes, mixed $at): array
    {
        return self::build(Rederivation::run($envelopes, $at));
    }

    /**
     * @param list<array<string,mixed>> $log
     * @return list<array<string,mixed>>
     */
    private static function groupings(array $log): array
    {
        $out = [];
        foreach ($log as $e) {
            if (($e['type'] ?? null) === Events::TYPE_CLAIM_ASSERTED && $e['kind'] === 'grouping') {
                $out[] = $e;
            }
        }

        return $out;
    }

    /**
     * @param array<string, list<mixed>> $keyToLegacy
     * @param array<string, array<string,mixed>> $perKey
     * @param array<string, array{value: mixed, codes: array<string, array{0: string, 1: string}>}> $entityCodes
     * @return array<string, array<string,mixed>>
     */
    private static function invert(array $keyToLegacy, array $perKey, array $entityCodes): array
    {
        // entity => list of keys.
        $entityToKeys = [];
        $entityValue = [];
        foreach ($keyToLegacy as $key => $entities) {
            foreach ($entities as $e) {
                $ek = self::entityKey($e);
                $entityToKeys[$ek][] = $key;
                $entityValue[$ek] = $e;
            }
        }

        $out = [];
        foreach ($entityToKeys as $ek => $keys) {
            $all = array_values(array_unique($keys));
            usort($all, static fn (string $a, string $b): int => self::keyNum($a) <=> self::keyNum($b));

            $out[(string) $ek] = [
                'primary' => self::primary($all, $perKey),
                'all' => $all,
                'relation' => self::relation($all, $entityValue[$ek], $keyToLegacy, $entityCodes),
            ];
        }

        return $out;
    }

    /**
     * @param list<string> $all
     * @param mixed $entity
     * @param array<string, list<mixed>> $keyToLegacy
     * @param array<string, array{value: mixed, codes: array<string, array{0: string, 1: string}>}> $entityCodes
     * @return string|array{0: string, 1: mixed}|array{0: string, 1: mixed, 2: string}
     */
    private static function relation(array $all, mixed $entity, array $keyToLegacy, array $entityCodes): string|array
    {
        if (count($all) !== 1) {
            return 'split';
        }
        $key = $all[0];

        $others = [];
        foreach ($keyToLegacy[$key] as $co) {
            if (self::entityKey($co) !== self::entityKey($entity)) {
                $others[] = $co;
            }
        }
        usort($others, self::compareEntities(...));

        if ($others === []) {
            return 'stable';
        }

        return self::merged($entity, $others, $entityCodes);
    }

    /**
     * @param mixed $entity
     * @param list<mixed> $others
     * @param array<string, array{value: mixed, codes: array<string, array{0: string, 1: string}>}> $entityCodes
     * @return array{0: string, 1: list<mixed>}|array{0: string, 1: list<mixed>, 2: string}
     */
    private static function merged(mixed $entity, array $others, array $entityCodes): array
    {
        $mine = $entityCodes[self::entityKey($entity)]['codes'] ?? [];

        $bridge = [];
        foreach ($others as $other) {
            $theirs = $entityCodes[self::entityKey($other)]['codes'] ?? [];
            $bridge = Sets::union($bridge, Sets::intersection($mine, $theirs));
        }

        foreach (Sets::values($bridge) as $code) {
            if (CodeRegistry::nationalGrade($code[0])) {
                return ['merged', $others];
            }
        }

        return ['merged', $others, 'suspect'];
    }

    /**
     * @param list<string> $all
     * @param array<string, array<string,mixed>> $perKey
     */
    private static function primary(array $all, array $perKey): string
    {
        if (count($all) === 1) {
            return $all[0];
        }

        $best = $all[0];
        $bestScore = self::primaryScore($all[0], $perKey);
        foreach ($all as $key) {
            $score = self::primaryScore($key, $perKey);
            if ($score > $bestScore) {
                $bestScore = $score;
                $best = $key;
            }
        }

        return $best;
    }

    /**
     * @param array<string, array<string,mixed>> $perKey
     * @return array{0: int, 1: int, 2: int}
     */
    private static function primaryScore(string $key, array $perKey): array
    {
        $info = $perKey[$key];

        return [self::spineRank($info['codes']), count($info['sources']), -self::keyNum($key)];
    }

    /** @param array<string, array{0: string, 1: string}> $codes */
    private static function spineRank(array $codes): int
    {
        foreach (Sets::values($codes) as $c) {
            if ($c[0] === 'cnk') {
                return 2;
            }
        }
        foreach (Sets::values($codes) as $c) {
            if ($c[0] === 'gtin') {
                return 1;
            }
        }

        return 0;
    }

    private static function keyNum(string $key): int
    {
        return (int) substr($key, strrpos($key, '_') + 1);
    }

    private static function entityKey(mixed $entity): string
    {
        return (is_int($entity) ? 'i:' : 's:').$entity;
    }

    private static function compareEntities(mixed $a, mixed $b): int
    {
        if (is_int($a) && is_int($b)) {
            return $a <=> $b;
        }

        return strcmp((string) $a, (string) $b);
    }
}
